geometric_kernels/eigenfunctions.py

```diff
@@ -0,0 +1,187 @@
+class Eigenfunctions(abc.ABC):
+    r"""
+    Represents a set of eigenfunctions of an operator. Referred to as :math:`Phi = [\phi_i]_{i=0}^{M-1}`.
+    """
+
+    def weighted_outerproduct(
+        self, weights: TensorLike, X: TensorLike, X2: Optional[TensorLike] = None
+    ) -> TensorLike:
+        r"""
+        Computes :math:`\sum_{i=0}^{M-1} w_i \phi_i(x1) \phi_i(x2)`.
+
+        :param weights: [M, 1]
+        :param X: Inputs where to evaluate the eigenfunctions, shape = [N, D].
+        :param X2: Inputs where to evaluate the eigenfunctions, shape = [N2, D].
+            Default to None, in which X is used for X2.
+        :return: shape [N, N2]
+        """
+        Phi_X = self.__call__(X)  # [N, M]
+        if X2 is None:
+            Phi_X2 = Phi_X
+        else:
+            Phi_X2 = self.__call__(X2)  # [N2, M]
+
+        weights = tf.reshape(weights, (-1,))
+        return tf.einsum("ni,ki,i->nk", Phi_X, Phi_X2, weights)  # [N, N2]
+
+    def weighted_outerproduct_diag(self, weights: TensorLike, X: TensorLike) -> TensorLike:
+        r"""
+        Computes :math:`\sum_{i=0}^{M-1} w_i \phi_i(x) \phi_i(x)`. Corresponds to the
+        diagonal elements of `weighted_outproduct` but they can be calculated more
+        efficiently.
+
+        :param weights: [M, 1]
+        :param X: Inputs where to evaluate the eigenfunctions, shape = [N, D].
+        :return: shape [N,]
+        """
+        Phi_X = self.__call__(X)  # [N, L]
+        weights = tf.reshape(weights, (-1,))
+        return tf.einsum("ni,i->n", Phi_X ** 2, weights)  # [N,]
+
+    @abc.abstractmethod
+    def __call__(self, x: TensorLike) -> TensorLike:
+        raise NotImplementedError
+
+    @abc.abstractproperty
+    def num_eigenfunctions(self) -> int:
+        """Number of eigenfunctions, M"""
+        raise NotImplementedError
+
+
+class EigenfunctionWithAdditionTheorem(Eigenfunctions):
+    """
+    Eigenfunctions for which the sum over a level has a simpler expression.
+    In the case the weights over a level in the `weighted_outproduct` our identical
+    we can make use of this expression to simplify computations.
+
+    We assume there are `L` levels. The sum of the number of eigenfunctions per level should
+    equal the total amount of eigenfunctions.
+    """
+
+    def weighted_outerproduct(
+        self, weights: TensorLike, X: TensorLike, X2: Optional[TensorLike] = None
+    ) -> TensorLike:
+        r"""
+        Computes :math:`\sum w_i \phi_i(x1) \phi_i(x2)`.
+
+        :param weights: [L, 1]
+            .. note:
+                The length of `weights` is equal to the number of levels.
+                This is **not** the same as the number of eigenfunctions.
+
+        :param X: Inputs where to evaluate the eigenfunctions, shape = [N, D].
+        :param X2: Inputs where to evaluate the eigenfunctions, shape = [N2, D].
+            Default to None, in which X is used for X2.
+        :return: shape [N, N2]
+        """
+        if X2 is None:
```

Comment:
Couldn't we avoid code duplication by putting `if X2 is None: X2=X`?

Reply:
good catch